### PR TITLE
Add OpenStack Cinder CSI driver support

### PR DESCRIPTION
**Pull request by Fedosin**

### CODE REVIEW

File: enhancements/storage/csi-cinder-operator.md [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+- OpenStack is a key cloud provider for OpenShift and use of Cinder has been supported in both OpenShift 3.x and 4.x, we need to move this driver to be CSI, provided by an operator.
+
+- CSI drivers (including OpenStack Cinder) will enable new storage features, not possible with in-tree volume plugins, like snapshots and volume cloning.
+
+### Goals
+
+- Create a cluster operator to install the OpenStack Cinder CSI driver.
+
+- Publish the operator on OperatorHub.
```
Comment (jsafrane) [inline enhancements/storage/csi-cinder-operator.md:46]: We don't want the operator on OperatorHub, it should be installed by CSO/CVO.

File: enhancements/storage/csi-cinder-operator.md [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+- Implement CSI tooling (snapshotter, resizer, cloner). It is available in OpenShift since 4.3.
+
+## Proposal
+
+OCP ships a new cluster operator called cinder-csi-driver-operator.
+
+1. The operator installs the OpenStack Cinder CSI driver following the [recommended guidelines (currently work-in-progress)](https://github.com/openshift/enhancements/pull/139/files).
+2. The operator deploys the all the objects required by the OpenStack Cinder CSI driver:
+   2.1 A namespace called `openshift-cinder-csi-driver`.
```
Comment (jsafrane) [inline enhancements/storage/csi-cinder-operator.md:62]: Use `openshift-cluster-csi-drivers` namespace, where all OCP CSI drivers should run (except for Manila)

File: enhancements/storage/csi-cinder-operator.md [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+   2.2 Two ServiceAccounts: one for the Controller Service and other for the Node Service of the CSI driver.
+   2.3 The RBAC rules to be used by the sidecar containers.
+   2.4 The CSIDriver object representing the CSI driver.
+   2.5 A Deployment that runs the driver's Controller Service.
+   2.7 A DaemonSet that runs the driver's Node Service.
+   2.8 A non-default StorageClass that uses the CSI driver as a provisioner.
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply enhancements/storage/csi-cinder-operator.md:68]: Cluster admin can choose if they're going to use stable & supported in-tree Cinder  volume plugin or tech preview CSI (and get snapshots). In-tree will be the default one for now.

File: enhancements/storage/csi-cinder-operator.md [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+3. The operator deploys all the CSI driver objects in the namespace `openshift-cinder-csi-driver`.
```
Comment (jsafrane) [inline enhancements/storage/csi-cinder-operator.md:69]: openshift-cluster-csi-drivers

File: enhancements/storage/csi-cinder-operator.md [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+   3.1 This is true regardless the namespace where the operator itself is deployed.
+4. The operator itself is installed by [Cluster Storage Operator](https://github.com/openshift/cluster-storage-operator) for all OpenStack clouds.
+   4.1 The operator should be deployed in the namespace `openshift-cluster-csi-drivers`.
+   4.2 The CR which the operator reacts to is *non-namespaced* and is named `cinder.csi.openstack.org`.
+   4.3 The CR belongs to `ClusterCSIDriver` CRD.
+5. The operator leverages the OpenStack credentials by creating a credentials request to [Cluster Credential Operator](https://github.com/openshift/cloud-credential-operator).
```
Comment (jsafrane) [inline enhancements/storage/csi-cinder-operator.md:75]: Should there be a note about self-signed OpenStack certificates?

File: enhancements/storage/csi-cinder-operator.md [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+#### Test the driver manually (Done)
+
+When all required components are built, we can manually deploy the driver and test how it works.
+
+Actions:
+
+- Manually deploy the driver on OCP on an OpenStack cloud.
+
+- Go through the lifecycle of volume (create, update, mount, write data, delete, etc.)
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply enhancements/storage/csi-cinder-operator.md:98]: +1, what can be covered by CI and what must be tested manually? Especially self-signed certificates and availability zones, do we / can we have them in CI?

File: enhancements/storage/csi-cinder-operator.md [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+#### Write OpenStack Cinder CSI driver operator
+
+The operator should be able to create, configure and manage OpenStack Cinder CSI driver for OpenShift. In other words, automate what has been done on the previous step.
+
+Actions:
+
+- Create a new repo in OpenShift’s github: github.com/openshift/cinder-csi-driver-operator
```
Comment (jsafrane) [inline enhancements/storage/csi-cinder-operator.md:106]: openstack-cinder-csi-driver-operator

File: enhancements/storage/csi-cinder-operator.md [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+### Risks and Mitigations
+
+#### OpenStack Cinder CSI driver doesn’t work properly
+
+The driver has not been tested on either OSP or OCP. The team uses Devstack + K8s plugin for the development.
+
+Severity: medium-high
+Likelihood: probably
+
+When testing on industrial clouds, we may face challenges, like bugs, issues with compatibility and so on. We will need to address them accordingly.
```
[the remark this answers is not in the document]
Comment (Fedosin) [reply enhancements/storage/csi-cinder-operator.md:156]: Done :)

File: enhancements/storage/csi-cinder-operator.md [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+
+### Test Plan
+
+- Both upstream and downstream CSI driver repositories will have a pre-submit job to run the Kubernetes external storage tests.
```
[the remark this answers is not in the document]
Comment (Fedosin) [reply enhancements/storage/csi-cinder-operator.md:160]: Upstream `quay.io/openshift`
Downstream`registry.redhat.io/openshift4`

File: enhancements/storage/csi-cinder-operator.md [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+
+- The operator will have unit tests and e2e tests, also running as pre-submit jobs.
+
+- The operator will also have a pre-submit job to run the Kubernetes external storage tests against the CSI driver installed by it.
```
[the remark this answers is not in the document]
Comment (Fedosin) [reply enhancements/storage/csi-cinder-operator.md:164]: agree, the operator won't provide e2e tests. we are going to reuse the existing csi test suite for the e2e job

File: enhancements/storage/csi-cinder-operator.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (Fedosin) [reply enhancements/storage/csi-cinder-operator.md:98]: unfortunately no, our CI supports neither self-signed certificates nor availability zones. we will have to test them manually, like we do with all other components.

File: enhancements/storage/csi-cinder-operator.md (hunk 5 above)
Comment (Fedosin) [reply enhancements/storage/csi-cinder-operator.md:75]: I don't think so. Unlike Manila, Cinder operator doesn't communicate with the underlying OpenStack cloud, so SSC support is not required. For the driver itself we just fetch certs from the config map and mount to the pods as usual.

Comment (jsafrane) [reply enhancements/storage/csi-cinder-operator.md]: Still, it must be tested by QA.

**Pull request merged**